### PR TITLE
Warehouse

### PR DESCRIPTION
**Pull request by pospieszynski**
MVP
to run: ruby sinatra.rb
#edit: changed to direct command: bundle exec rackup 
@gotar

### CODE REVIEW

**Comment by gotar** [comment]
to run: bundle exec rackup

File: lib/services/add_commodity_to_warehouse.rb [hunk 1]
```diff
@@ -1,18 +1,17 @@
 require_relative './fetch_product'
 require_relative './find_product_in_warehouse_by_id'
 
-
 module Shop
   class AddCommodityToWarehouse
-    def call(id, quantity)
+    def call(id, quantity = 1)
       product = FetchProduct.new().call(id)
```
Comment (gotar) [inline lib/services/add_commodity_to_warehouse.rb]: do not use `()` if no argument passed

File: lib/services/add_commodity_to_warehouse.rb [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
       return unless product
       commodity = FindProductInWarehouseById.new().call(id)
```
Comment (gotar) [inline lib/services/add_commodity_to_warehouse.rb]: the same, do not use `()`

File: lib/services/add_product_to_basket.rb [hunk 3]
```diff
@@ -2,15 +2,11 @@
...
-
   class AddProductToBasket
-
     def call(id, quantity = 1)
-
       commodity = FindProductInWarehouseById.new().call(id)
-      return unless  commodity
+      return unless commodity
       return unless commodity.quantity >= quantity
       Shop::DeleteCommodityFromWarehouse.new().call(id, quantity)
```
Comment (gotar) [inline lib/services/add_product_to_basket.rb]: drop empty `()`

File: lib/services/add_product_to_basket.rb [hunk 4]
```diff
@@ -2,15 +2,11 @@
...
       if commodity = Shop::FindProductInBasket.new().call(id)
```
Comment (gotar) [inline lib/services/add_product_to_basket.rb]: drop empty ()

File: lib/services/delete_commodity_from_warehouse.rb [hunk 5]
```diff
@@ -4,7 +4,7 @@ module Shop
   class DeleteCommodityFromWarehouse
     def call(id, quantity = 1)
```
Comment (gotar) [inline lib/services/delete_commodity_from_warehouse.rb]: better to use more descriptive name, in place of ID use proper name, like product_id

File: lib/services/delete_commodity_from_warehouse.rb [hunk 6]
```diff
@@ -4,7 +4,7 @@ module Shop
...
       commodity = FetchCommodity.new().call(id)
-
+      puts "COMMODITY" + commodity.inspect
```
Comment (gotar) [inline lib/services/delete_commodity_from_warehouse.rb]: do not puts anything in code

File: lib/services/fetch_commodity.rb [hunk 7]
```diff
@@ -1,9 +1,7 @@
 module Shop
   class FetchCommodity
     def call(id)
```
Comment (gotar) [inline lib/services/fetch_commodity.rb]: change `id` to something more descriptive

File: lib/services/fetch_product.rb [hunk 8]
```diff
@@ -1,9 +1,7 @@
 module Shop
   class FetchProduct
     def call(id)
```
Comment (gotar) [inline lib/services/fetch_product.rb]: change ID to something more descriptive

File: lib/services/find_product_in_warehouse_by_id.rb [hunk 9]
```diff
@@ -1,9 +1,7 @@
 module Shop
   class FindProductInWarehouseById
     def call(id)
-      WAREHOUSE.find do |commodity|
-        commodity.product_id == id
-      end
+      WAREHOUSE.find{ |commodity| commodity.product_id == id }
```
Comment (gotar) [inline lib/services/find_product_in_warehouse_by_id.rb]: change ID to something more descriptive

File: lib/services/price_net.rb [hunk 10]
```diff
@@ -1,10 +1,8 @@
 module Shop
-
   class PriceNet
-
     def call
       puts WAREHOUSE.inspect
-      BASKET.reduce(0){ |sum, n| sum += n.quantity * PRODUCTS_CATALOGUE.find{ |product| product.id == n.product_id }.price }.to_f.round(2)
+      BASKET.reduce(0){ |sum, n| sum += n.quantity * PRODUCTS_CATALOGUE.find { |product| product.id == n.product_id }.price }.to_f.round(2)
```
Comment (gotar) [inline lib/services/price_net.rb]: split it to more methods, hard to read if everything in one line

File: lib/services/receipt.rb [hunk 11]
```diff
@@ -3,17 +3,12 @@
...
     def call
-
       receipt = "\n*******************"
       receipt += "\n*BIEDRONKA*"
       receipt += "\n*******************"
 
-
-
       BASKET.each do |commodity|
         product = FetchProduct.new().call(commodity.product_id)
```
Comment (gotar) [inline lib/services/receipt.rb]: drop ()

File: sinatra.rb [hunk 12]
```diff
@@ -1,46 +1,59 @@
...
 
 get "/products" do
-  @products = []
-  Shop::PRODUCTS_CATALOGUE.each do |product|
-      # product = Shop::PRODUCTS_CATALOGUE.find{ |product| product.id == commodity.product_id }
-      @products << product
+  products = []
+  Shop::PRODUCTS_CATALOGUE.each do |product| #replace with fetch porducts service
+      products << product
   end
```
Comment (gotar) [inline sinatra.rb]: move to service

File: sinatra.rb [hunk 13]
```diff
@@ -1,46 +1,59 @@
...
+  products = []
   Shop::BASKET.each do |commodity|
-    @products << Shop::PRODUCTS_CATALOGUE.find{ |product| product.id == commodity.product_id }.to_hash.merge!( { quantity: commodity.quantity } )
+    products << Shop::PRODUCTS_CATALOGUE.find{ |product| product.id == commodity.product_id }.to_hash.merge!( { quantity: commodity.quantity } )
   end
-  @gross = Shop::PriceGross.new().call
-  @net = Shop::PriceNet.new().call
-  erb :basket
+  gross = Shop::PriceGross.new().call
+  net = Shop::PriceNet.new().call
```
Comment (gotar) [inline sinatra.rb]: drop ()

File: sinatra.rb [hunk 14]
```diff
@@ -1,46 +1,59 @@
...
+  erb :"basket/index", locals: {gross: gross, net: net, products: products }
 end
 
-post '/basket/add' do
+get '/warehouse' do
+  products = []
+  Shop::WAREHOUSE.each do |commodity|
+    products << Shop::PRODUCTS_CATALOGUE.find{ |product| product.id == commodity.product_id }.to_hash.merge!( { quantity: commodity.quantity } )
```
Comment (gotar) [inline sinatra.rb]: too long line

File: sinatra.rb [hunk 15]
```diff
@@ -1,46 +1,59 @@
...
+  end
+  catalogue = Shop::FetchProducts.new().call
```
Comment (gotar) [inline sinatra.rb]: drop ()

File: sinatra.rb [hunk 16]
```diff
@@ -1,46 +1,59 @@
...
+  erb :"warehouse/index", locals: {catalogue: catalogue, products: products }
+end
+
+delete '/warehouse' do
+  Shop::DeleteCommodityFromWarehouse.new().call( params['product_id'].to_i )
```
Comment (gotar) [inline sinatra.rb]: drop ()

File: sinatra.rb [hunk 17]
```diff
@@ -1,46 +1,59 @@
...
+  redirect '/warehouse'
+end
+
+post '/warehouse' do
+  Shop::AddCommodityToWarehouse.new().call( params['product_id'].to_i, params['quantity'].to_i )
```
Comment (gotar) [inline sinatra.rb]: drop ()

File: sinatra.rb [hunk 18]
```diff
@@ -1,46 +1,59 @@
...
+  redirect '/warehouse'
+end
+
+post '/basket' do
   Shop::AddProductToBasket.new().call( params['product_id'].to_i, params['quantity'].to_i  )
```
Comment (gotar) [inline sinatra.rb]: drop ()

File: sinatra.rb [hunk 19]
```diff
@@ -1,46 +1,59 @@
...
   redirect '/basket'
 end
 
 delete '/basket' do
-  puts params['product_id'].to_i
   Shop::RemoveProductFromBasket.new().call( params['product_id'].to_i )
```
Comment (gotar) [inline sinatra.rb]: drop ()

File: views/basket/index.erb [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    <th>Quantity</th>
+    <th>Price</th>
+    <th>Action</th>
+  </tr>
+
+  <%products.each do |product| %>
+  <tr>
+    <td><%= product[:name] %></td>
+    <td><%= product[:quantity] %></td>
+    <td><%= "%0.2f" %  (product[:price] * product[:quantity] ) %>  PLN </td>
```
Comment (gotar) [inline views/basket/index.erb]: move to model (those calculation)

File: views/basket/index.erb [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    <td>
+    <form action="/basket" method="post">
+      <input name="_method" type="hidden" value="delete" />
+      <input type="hidden" name="product_id" value="<%= product[:id] %>">
+      <input type="submit" value="delete">
+    </form>
```
Comment (gotar) [inline views/basket/index.erb]: wrong indentation

File: views/products/show.erb [hunk 22]
```diff
@@ -0,0 +1,14 @@
+<h2> <%= product.name %></h2>
+<p> ID: <%= product.id %></p>
+<p> <%= product.price %>  PLN</p>
+<% if commodity %>
+  <p> <%= commodity.quantity %>  szt. available on the warehouse</p>
+  <form action="/basket" method="post">
+    <input type="hidden" name="product_id" value="<%= product.id %>">
+    <label for="qt">Quantity: </label>
```
Comment (gotar) [inline views/products/show.erb]: why qt? not quantity. Do not afraid to use long var names

File: lib/models/total_net.rb [hunk 23]
```diff
@@ -0,0 +1,8 @@
+module Shop
+  class TotalNet
+    attr_reader :value
+    def initialize price, quantity
```
Comment (gotar) [inline lib/models/total_net.rb]: add empty line between attr_reader and def

File: lib/services/fetch_commodity.rb [hunk 24]
```diff
@@ -1,9 +1,7 @@
 module Shop
   class FetchCommodity
-    def call(id)
-      WAREHOUSE.find do |commodity|
-        commodity.id == id
-      end
+    def call( product_id )
```
Comment (gotar) [inline lib/services/fetch_commodity.rb]: stop white characters (spaces) around product_id

File: lib/services/price_gross.rb [hunk 25]
```diff
@@ -3,7 +3,7 @@
 module Shop
   class PriceGross
     def call
-      PriceNet.new().call() * 1.23
+      PriceNet.new.call() * 1.23
```
Comment (gotar) [inline lib/services/price_gross.rb]: drop () after call

File: lib/services/price_net.rb [hunk 26]
```diff
@@ -1,10 +1,13 @@
 module Shop
-
   class PriceNet
-
     def call
-      puts WAREHOUSE.inspect
-      BASKET.reduce(0){ |sum, n| sum += n.quantity * PRODUCTS_CATALOGUE.find{ |product| product.id == n.product_id }.price }.to_f.round(2)
+      price = BASKET.reduce(0){ |sum, n| sum += n.quantity * PRODUCTS_CATALOGUE.find { |product| product.id == n.product_id }.price }
```
Comment (gotar) [inline lib/services/price_net.rb]: too long line

File: shop.rb [hunk 27]
```diff
@@ -1,77 +1,64 @@
...
-    Product.new('Bread', 3),
-    Product.new('Soap', 4),
-    Product.new('Trousers', 20),
-    Product.new('Pizza', 6),
+      Product.new('Milk', 2),
+      Product.new('Bread', 3),
+      Product.new('Soap', 4),
+      Product.new('Trousers', 20),
+      Product.new('Pizza', 6),
   ]
```
Comment (gotar) [inline shop.rb]: wrong indentation

File: shop.rb [hunk 28]
```diff
@@ -1,77 +1,64 @@
...
-    Commodity.new(2,2),
-    Commodity.new(3,2),
-    Commodity.new(4,2),
-    Commodity.new(5,2),
+      Commodity.new(1, 2),
+      Commodity.new(2, 2),
+      Commodity.new(3, 2),
+      Commodity.new(4, 2),
+      Commodity.new(5, 2),
   ]
```
Comment (gotar) [inline shop.rb]: wrong indentation

File: shop.rb [hunk 29]
```diff
@@ -1,77 +1,64 @@
...
-  #   def self.run
-  #     PRODUCTS_CATALOGUE.each do |product|
-  #       #CreateWarehouse.new.call(product.id, rand(1..10))
-  #     end
-  #   end
-  # end
+  class Main
+    def self.run
+      PRODUCTS_CATALOGUE.each do |product|
+        #CreateWarehouse.new.call(product.id, rand(1..10))
```
Comment (gotar) [inline shop.rb]: do not keep commented code in file, delete it

File: sinatra_app.rb [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+      Product.new('Pizza', 6),
+  ]
+
+  WAREHOUSE = [
+      Commodity.new(1, 2),
+      Commodity.new(2, 2),
+      Commodity.new(3, 2),
+      Commodity.new(4, 2),
+      Commodity.new(5, 2),
+  ]
```
Comment (gotar) [inline sinatra_app.rb]: wrong indentation

File: sinatra_app.rb [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+
+module Shop
+
+  PRODUCTS_CATALOGUE = [
+      Product.new('Milk', 2),
+      Product.new('Bread', 3),
+      Product.new('Soap', 4),
+      Product.new('Trousers', 20),
+      Product.new('Pizza', 6),
+  ]
```
Comment (gotar) [inline sinatra_app.rb]: wrong indentation

File: sinatra_app.rb [hunk 32]
```diff
@@ -0,0 +1,88 @@
...
+      Commodity.new(2, 2),
+      Commodity.new(3, 2),
+      Commodity.new(4, 2),
+      Commodity.new(5, 2),
+  ]
+
+  BASKET = []
+
+  class App < Sinatra::Base
+
```
Comment (gotar) [inline sinatra_app.rb]: drop empty line

File: views/basket/index.erb [hunk 33]
```diff
@@ -0,0 +1,32 @@
+<h3>Basket</h3>
+
+<table border="1">
+  <tr>
+    <th>Product</th>
+    <th>Quantity</th>
+    <th>Price</th>
+    <th>Action</th>
+  </tr>
+
```
Comment (gotar) [inline views/basket/index.erb]: drop empty line

File: views/basket/index.erb [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+        </td>
+      </tr>
+  <% end %>
+  <tr>
+    <td colspan="4">Total gross: <%= "%0.2f" % gross %>  PLN</td>
+  </tr>
+  <tr>
+    <td colspan="4">Total net: <%= "%0.2f" % net %>  PLN</td>
+  </tr>
+
```
Comment (gotar) [inline views/basket/index.erb]: drop empty line

File: views/basket/index.erb [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+        <td><%= product[:quantity] %></td>
+        <td><%= product[:total_gross] %>  PLN</td>
+        <td>
+          <form action="/basket_remove" method="post">
+            <input name="_method" type="hidden" value="delete"/>
+            <input type="hidden" name="product_id" value="<%= product[:id] %>">
+            <input type="submit" value="delete">
+          </form>
+        </td>
+      </tr>
```
Comment (gotar) [inline views/basket/index.erb]: wrong indentation

File: views/products/index.erb [hunk 36]
```diff
@@ -0,0 +1,14 @@
...
+  <tr>
+    <th>Product</th>
+    <th>Price</th>
+  </tr>
+  <% products.each do |product| %>
+      <tr>
+        <td><a href="/products/<%= product.id %>"> <%= product.name %></a></td>
+        <td><%= product.price %> PLN</td>
+      </tr>
+  <% end %>
```
Comment (gotar) [inline views/products/index.erb]: wrong indentation

File: views/products/show.erb [hunk 37]
```diff
@@ -0,0 +1,14 @@
...
+    <p> <%= commodity.quantity %>  szt. available on the warehouse</p>
+    <form action="/basket" method="post">
+      <input type="hidden" name="product_id" value="<%= product.id %>">
+      <label for="quantity">Quantity: </label>
+      <input type="number" min="1" max="<%= commodity.quantity %>" step="1" value="1" name="quantity" id="quantity">
+      <input type="submit" value="Add to basket">
+    </form>
+<% else %>
+    <p>Product not available</p>
+<% end %>
```
Comment (gotar) [inline views/products/show.erb]: wrong indentation

File: views/warehouse/index.erb [hunk 38]
```diff
@@ -0,0 +1,37 @@
...
+    <td>
+      <form action="/warehouse_remove" method="post">
+        <input name="_method" type="hidden" value="delete" />
+        <input type="hidden" name="product_id" value="<%= product[:id] %>">
+        <input type="submit" value="delete">
+      </form>
+    </td>
+  </tr>
+  <% end %>
+
```
Comment (gotar) [inline views/warehouse/index.erb]: drop empty line

File: views/warehouse/index.erb [hunk 39]
```diff
@@ -0,0 +1,37 @@
...
+</table>
+</br>
+
+<form action="/warehouse" method="post" id="product">
+  <select name="product_id" form="product">
+    <%catalogue.each do |product| %>
+        <option value="<%= product.id %>"><%= product.name %></option>
+    <%end%>
```
Comment (gotar) [inline views/warehouse/index.erb]: wrong indentation

Comment (gotar) [inline views/warehouse/index.erb]: add spaces around `end`

File: .gitignore [hunk 40]
```diff
@@ -0,0 +1 @@
+/coverage/
```
Comment (gotar) [inline .gitignore]: change to coverage

File: coverage/.resultset.json [hunk 41]
```diff
@@ -0,0 +1,427 @@
+{
```
Comment (gotar) [inline coverage/.resultset.json]: drop all coverge files!!!

Comment (gotar) [inline coverage/.resultset.json]: check what file you add to commit twice before you push

File: lib/services/price_net.rb [hunk 42]
```diff
@@ -1,10 +1,13 @@
...
-
     def call
-      puts WAREHOUSE.inspect
-      BASKET.reduce(0){ |sum, n| sum += n.quantity * PRODUCTS_CATALOGUE.find{ |product| product.id == n.product_id }.price }.to_f.round(2)
+      price = BASKET.reduce(0){ |sum, n| sum += n.quantity * FetchProduct.new.call(n.product_id).price }
+      price = convert_to_fload_and_round price
+    end
+
+    private
+    def convert_to_fload_and_round price
```
Comment (gotar) [inline lib/services/price_net.rb]: you should add () around price. use always () when you have any var

File: public/application.css [hunk 43]
```diff
@@ -0,0 +1,3 @@
+body{
+    background-color: red;
+}
```
Comment (gotar) [inline public/application.css]: add empty line

File: views/warehouse/index.erb [hunk 44]
```diff
@@ -0,0 +1,33 @@
...
+        <input type="submit" value="delete">
+      </form>
+    </td>
+  </tr>
+  <% end %>
+</table>
+</br>
+<form action="/warehouse" method="post" id="product">
+  <select name="product_id" form="product">
+    <%catalogue.each do |product| %>
```
Comment (gotar) [inline views/warehouse/index.erb]: always use space after and before `%`

File: lib/services/price_net.rb (hunk 42 above)
Comment (pospieszynski) [inline lib/services/price_net.rb]: W przypadku testów jest inna konwencja? Bo z tego co zaobserwowałem, po wywołaniu metody "to" nie używałes nawiasów:
expect( # ).to paramater

File: lib/models/total_net.rb [hunk 45]
```diff
@@ -0,0 +1,9 @@
+module Shop
+  class TotalNet
+    attr_reader :value
+
+    def initialize price, quantity
+      @value = price * quantity
+    end
+  end
+end
```
Comment (gotar) [inline lib/models/total_net.rb]: add empty line (or better config your editor to add it automatically)

**Comment by gotar** [comment]
good enough, you can merge

**Pull request merged**